### PR TITLE
Doc mesh and tube mesh generation 

### PR DESCRIPTION
**Pull request by kerautret**
Module doc add with some static method to generate tube mesh and height field mesh

### CODE REVIEW

File: examples/io/display3DToOFF.cpp [hunk 1]
```diff
@@ -1,3 +1,8 @@
+
+
+
+
+
```
Comment (dcoeurjo) [inline examples/io/display3DToOFF.cpp]: remove ? ;)

File: ChangeLog.md [hunk 2]
```diff
@@ -52,7 +52,8 @@
    ConstIterator and adds a new method to change the color of a
    specific face. (Bertrand Kerautret,
    [#937](https://github.com/DGtal-team/DGtal/pull/937))
-
+ - New methods to generate basic mesh of 3D tube and height field. New mesh module documentation.
```
Comment (dcoeurjo) [inline ChangeLog.md]: New methods to generate basic 3D tubular meshes and height fields. New mesh module documentation added. 

?

File: src/DGtal/shapes/Mesh.h [hunk 3]
```diff
@@ -386,13 +388,56 @@ namespace DGtal
...
-  protected:
 
   
+    // ------------------------- Mesh generation ------------------------------
+
+  public:
+    
+    /**
+     * Generates a tube mesh from a set of points representing the
+     * tube skeleton.  Each circular sections are connected with
```
Comment (dcoeurjo) [inline src/DGtal/shapes/Mesh.h]: each circular section is ...

File: src/DGtal/shapes/Mesh.h [hunk 4]
```diff
@@ -386,13 +388,56 @@ namespace DGtal
...
+    static
+    void
+    createTubularMesh(Mesh<TPoint> &aMesh,  const std::vector<TPoint> &aSkeleton, const double aRadius, 
+                      const double angleStep = 0.2, const DGtal::Color &aMeshColor = DGtal::Color::White );
+
 
 
-          
+    /**
+     * Generates a surface mesh defined from an sequence of a 2D
```
Comment (dcoeurjo) [inline src/DGtal/shapes/Mesh.h]: a sequence

Comment (dcoeurjo) [inline src/DGtal/shapes/Mesh.h]: of 2D height values

File: src/DGtal/shapes/Mesh.ih [hunk 5]
```diff
@@ -289,6 +289,139 @@ DGtal::Mesh<TPoint>::className() const
...
+
+  // Generating vertices..
+  for(unsigned int i = 0; i< aSkeleton.size(); i++)
+    {
+      TPoint vectDir;
+      TPoint uDir1, uDirPrec;
+      TPoint uDir2;
+      TPoint firstPoint;
+      
+      if(i != aSkeleton.size()-1){
```
Comment (dcoeurjo) [inline src/DGtal/shapes/Mesh.ih]: Please use  C ANSI style with brackets "{"  in a new line.

File: src/DGtal/shapes/Mesh.ih [hunk 6]
```diff
@@ -289,6 +289,139 @@ DGtal::Mesh<TPoint>::className() const
...
+      double d = -vectDir[0]* aSkeleton.at(i)[0] - vectDir[1]*aSkeleton.at(i)[1] - vectDir[2]*aSkeleton.at(i)[2];
+      TPoint pRefOrigin;
+        if(vectDir[0]!=0){
+          pRefOrigin [0]= -d/vectDir[0];
+          pRefOrigin [1]= 0.0;
+          pRefOrigin [2]= 0.0;
+          if(pRefOrigin==aSkeleton.at(i)){
+            pRefOrigin[1]=-1.0;
+          }
+        }else if (vectDir[1]!=0){
```
Comment (dcoeurjo) [inline src/DGtal/shapes/Mesh.ih]: else , if and { should be in separate new lines)

File: src/DGtal/shapes/doc/moduleMesh.dox [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+@page moduleMesh Manipulating Simple Mesh 
+@author Bertrand Kerautret
+
+[TOC]
+
+\section sectmoduleMesh1 Simple Mesh Class 
+
+
+The aim of this class is to handle basic simple mesh mainly for the
+task of import/export/display of digital surface. This class is able
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleMesh.dox]: for basic tasks such as to import/export and display digital surfaces

File: src/DGtal/shapes/doc/moduleMesh.dox [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+to deal with triangulated, quad or polygonal faces and can include
+specific colors to display faces. It also provides several static
+tools to generate simple shapes like tubular objects or the 3d surface
+from height map.
+
+The mesh object store explicitly each vertex and each face are represented with the list of point index.
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleMesh.dox]: stores explicitly the vertices and each face is represented as a list of vertex indices

File: src/DGtal/shapes/doc/moduleMesh.dox [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+\subsection subsect2moduleMesh Mesh Construction
+
+To construct a Mesh object you need first to include the associated
+header file:
+
+@snippet examples/shapes/mesh3DConstructionAndVisualisation.cpp MeshUseInclude 
+
+
+Then the mesh construction is very simple, for instance you can start
+to construct an empty mesh input and specify if you want to store or
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleMesh.dox]: start construcing ... ?

File: src/DGtal/shapes/doc/moduleMesh.dox [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+not the vertex colors in the mesh:
+
+@snippet examples/shapes/mesh3DConstructionAndVisualisation.cpp MeshUseMeshConstructor 
+
+Then you can add some vertex mesh:
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleMesh.dox]: some vertices to the mesh ?

File: src/DGtal/shapes/doc/moduleMesh.dox [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+@snippet examples/shapes/mesh3DConstructionAndVisualisation.cpp MeshUseMeshAddingPoints 
+
+
+and add  some triangular/quad faces from the vertex index:
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleMesh.dox]: from vertex indices

File: src/DGtal/shapes/doc/moduleMesh.dox [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+@snippet examples/shapes/mesh3DConstructionAndVisualisation.cpp MeshUseMeshAddingPolygonalFaces 
+
+An example of mesh construction is given here:  mesh3DConstructionAndVisualisation.cpp and will produce the following faces:
+
+   \image html  example3dMeshConstruction.png  "Simple example of 3D mesh construction."
+
+
+\subsection subsect2moduleMeshIO Mesh IO
+
+The Mesh object are integrated in the main IO framework of DGtal. You
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleMesh.dox]: is integrated (or objectS)

File: src/DGtal/shapes/doc/moduleMesh.dox [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+@snippet tests/shapes/testMesh.cpp testMeshCreateTubeMesh
+
+You will obtain such a tube display:
+   \image html  testMeshVisu3d.png  "Resulting mesh of tubular object."
+   \image latex testMeshVisu3d.png  "Resulting mesh of tubular object."  width=5cm
+
+
+
+Another example is given in testMesh.cpp which generates an height
+field from a sequence of height. In the same way than the previous
```
Comment (dcoeurjo) [inline src/DGtal/shapes/doc/moduleMesh.dox]: of height values ?

**Comment by kerautret** [comment]
Thanks for all remarks that I agree (and sorry for the various typos).
I also add some edits on iterators and the previous commit 9dfc265 also contains some edits on (const/not const getVertex)

**Comment by kerautret** [comment]
To finish, I also added an new method to display tube with variable radius...

**Comment by dcoeurjo** [comment]
I'll have a look to the build . thanks for the edits

**Comment by dcoeurjo** [comment]
Build and test are fine. thanks.

**Pull request merged**